backends/ebpf/psa/ebpfPsaTable.cpp

```diff
@@ -86,6 +87,41 @@ EBPFTablePSA::EBPFTablePSA(const EBPFProgram* program, const IR::TableBlock* tab
         }
         this->size = 1;
     }
+
+    initImplementation();
+}
+
+EBPFTablePSA::EBPFTablePSA(const EBPFProgram* program, CodeGenInspector* codeGen, cstring name) :
+                           EBPFTable(program, codeGen, name), implementation(nullptr) {}
+
+void EBPFTablePSA::initImplementation() {
+    // PSA table is allowed to have up to one table implementation. Function forEachPropertyEntry
+    // will iterate over all entries in property, so lets use of this and print errors.
+    auto impl = [this](const IR::PathExpression * pe) {
+        CHECK_NULL(pe);
+        auto decl = program->refMap->getDeclaration(pe->path, true);
+        auto di = decl->to<IR::Declaration_Instance>();
+        CHECK_NULL(di);
+        cstring type = di->type->toString();
+
+        if (this->implementation != nullptr) {
+            ::error(ErrorType::ERR_UNSUPPORTED,
+                    "%1%: Up to one implementation is supported in a table", pe);
+            return;
+        }
+
+        if (type == "ActionProfile") {
+            auto ap = program->control->getTable(di->name.name);
+            this->implementation = ap->to<EBPFTableImplementationPSA>();
+        }
+
+        if (this->implementation != nullptr)
+            this->implementation->registerTable(this);
+        else
+            ::error(ErrorType::ERR_UNKNOWN,
+                    "%1%: unknown table implementation %2%", pe, decl);
+    };
+    forEachPropertyEntry("psa_implementation", impl);
```

Comment:
Earlier it might be hard to do this because this code is executed when creating table object (function is called by constructor), but I was able to rewrite it with standard visitors.